mantidproject/mantid
# Elemental Analysis - Run Data Loading

**Description of work.**
Adds a load widget consisting of: a `spinbox` with the ability to change runs; a `User Dirs` button to allow for the inclusion of run-containing folders; a `Load` button to load the selected run; a `Co-Add` button to allow the addition of runs.

The original Periodic Table functionality has *not* changed.

**Report to:** Aidy

---
**Functionality**
- `Load` and `Co-Add` operate on the number in the `spinbox`
- `Load` and `Co-Add` do not work on runs that don't exist (within the user directories)
- `Load` will reset the list of runs to be co-added, even if the run has already been loaded
-  `Co-Add` will not co-add the same run twice
- `User Dirs` will bring up the standard `Manage User Directories` GUI.
- When (co-)loaded, runs should be in the following tree format (where the order of the actual workspaces within the same group is irrelevant):
```
Run
    Run; Detector 1
        1_Total_Run
        1_Prompt_Run
        1_Delayed_Run
    Run; Detector 2
        2_Total_Run
        ...
```
- When co-added, subsequent runs are grouped into a range with `-`, whereas other runs are separated by `,`; the overall string is in numeric order: e.g.
    - Runs: `1, 2, 3, 4, 5` -> `1-5`
    - Runs: `1, 3, 5, 6, 7` -> `1, 3, 5-7`
    - Runs: `1, 3, 5` -> `1, 3, 5`
    - Runs: `1, 2, 4, 5, 7, 8` -> `1-2, 4-5, 7-8`
- For co-loading, the run numbers (as in the above tree diagram) are replaced with the `hyphenised` run string
---
**To test:**
- `Interfaces` -> `Muon` -> `Elemental Analysis`
- Extract the folder from [test_datafiles.zip](https://github.com/mantidproject/mantid/files/2283635/test_datafiles.zip) and add it to your user directories
- Sorting the workspaces by `Descending` and `Last Modified` might be helpful
- The runs contained are: `1234, 1235, 1236, 1237, 1238, 1239` (random data)
- The various functions, listed under `Functionality` can now be tested
- When co-adding, ensure that (for a given `x value`) the `y values` are being added correctly by viewing the data using mantid's data viewer


Closes #23023

*This does not require release notes* because **they were added in a previous PR**

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
